### PR TITLE
Add output sig_figs argument

### LINKED ISSUE #521: Truncation of large parameter values in cmdstan output
#### Summary:

Draws of parameter values from Cmdstan are truncated at 6 significant digits, resulting in loss of precision for parameters with values > 1e5 (like `lp__` in models with many observations).

#### Reproducible Steps:
Sample from this model in cmdstan:

```
data {
} 
parameters {
  real theta;
}
model {
  theta ~ normal(1e6, 1);
}
```

#### Current Output:

Draws for `theta` will be discrete, likely: `999997, 999998, 999999,  1e+06`.

#### Expected Output:

`theta - 1e6` should generate draws from a standard normal distribution. 

#### Current Version:
v2.13.1

### LINKED ISSUE #922: User-defined precision for CSV data
#### Summary:

The current default precision of CSV is fine for the majority of cases. There are however some cases where we want to override the default setting and output with more precision.  For example, if we want to output a generated covariance matrix.

The services layer leaves the precision as definied for the input ostream, if I read the code correctly. In this case Cmdstan could use `std::setprecision` on the ostream to set user-defined precision.

The argument would fall undert `ouput`

so we would call like

`./bernoulli sample data file=data.json output precision=12 file=out.csv`

#### Current Version:
v2.24.0

### PR DESCRIPTION
**Pull request by rok-cesnovar**
#### Submisison Checklist

- [x] Run tests: `./runCmdStanTests.py src/test`
- [x] Declare copyright holder and open-source license: see below

#### Summary:

Fixes #922 by adding the `output sig_figs` argument

Example call
`./examples/bernoulli/bernoulli sample data file=examples/bernoulli/bernoulli.data.json output sig_figs=11`

This also fixes #521 

For 
```
parameters {
  real theta;
}
model {
  theta ~ normal(1e6, 1);
}
```
we now get 

```
-0.163974438784448,0.996100733854735,0.923577280498764,2,7,0,0.405733542768386,1000000.57266821
-0.032493334636212,0.973398474520744,0.923577280498764,1,3,0,0.386216333045433,999999.745075169
-0.0468632717772731,0.981221965207712,0.923577280498764,2,7,0,0.1780758772537,1000000.30614791
-0.00119945149551241,0.972963915872715,0.923577280498764,2,3,0,0.2717442954076,1000000.0489786
-0.0564371215940085,0.985588929563464,0.923577280498764,1,3,0,0.100460796895455,999999.664032378
-1.24354467691627,0.88158579117078,0.923577280498764,2,3,0,1.24372474624881,1000001.57705084
-0.684138060691333,1,0.923577280498764,1,1,0,1.16139002990989,1000001.16973335
-0.22266328299244,0.860289977025518,0.923577280498764,2,7,0,1.88421272720085,999999.33267207
-0.0437246249548987,0.972169026825194,0.923577280498764,1,3,0,0.523107687642625,1000000.29571819
-0.0617620496505439,0.977932638436175,0.923577280498764,1,3,0,0.236767714319316,999999.648540046
-0.101074407933952,0.963661752494806,0.923577280498764,1,3,0,0.364584013491662,1000000.44960963
-0.274785595783059,0.91076712455456,0.923577280498764,1,3,0,0.823504712089934,999999.25866931
-0.437668291804487,0.969689929062161,0.923577280498764,1,1,0,0.454459792176465,999999.064405759
-0.921700893480246,0.912593622711634,0.923577280498764,1,1,0,0.92717810908897,999998.642280667
-0.647790804911401,0.937919559743682,0.923577280498764,2,3,0,1.76245720192101,999998.861763816
-0.0541615041274764,0.911815471508332,0.923577280498764,2,3,0,1.61248233940939,1000000.32912461
-0.00144424026582138,0.986496945235532,0.923577280498764,2,3,0,0.174918818217889,999999.946255414
-0.0280591161597656,0.99539065590599,0.923577280498764,1,3,0,0.0338569878478136,999999.763107129
```
instead of
```
-1.38713,0.847666,0.934761,2,3,0,1.47098,1e+06
-0.283844,1,0.934761,2,7,0,1.17849,999999
-0.792056,0.80529,0.934761,1,3,0,2.11822,1e+06
-0.997812,0.958009,0.934761,1,1,0,1.11126,1e+06
-1.13757,0.971282,0.934761,1,1,0,1.30806,1e+06
-0.0687194,0.987487,0.934761,1,3,0,1.06987,1e+06
-0.109481,0.963541,0.934761,1,3,0,0.357786,1e+06
-0.0644989,0.97546,0.934761,1,3,0,0.339564,1e+06
-0.120563,0.963662,0.934761,1,3,0,0.370343,1e+06
-0.106257,0.96239,0.934761,1,3,0,0.45239,1e+06
-0.0445222,0.978003,0.934761,1,3,0,0.286253,1e+06
-0.168583,0.946624,0.934761,2,3,0,0.474501,1e+06
-0.038764,0.96459,0.934761,2,3,0,0.496828,1e+06
-1.88103,0.689543,0.934761,1,3,0,2.31665,999998
-0.712226,1,0.934761,1,1,0,1.6397,999999
```

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Rok Češnovar, Uni. of Ljubljana

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by rok-cesnovar** [comment]
@seantalts mind looking at this one. I am tagging you as you threw light at this issue but let me know if you are busy. I can find someone else to review. Would like to get this one in for the release.

**Review by mitzimorris — COMMENTED** [review]
code looks good.
question about name "precision" - why not "sig_figs"?

two reasons:

- elsewhere we use "sig_figs", other pkgs use "digits" 
- "precision" as technical term is inverse variance - possibly confusing?

**Comment by mitzimorris** [comment]
agree we should get this in for release - note CmdStanPy issue:  https://github.com/stan-dev/cmdstanpy/issues/300

**Comment by rok-cesnovar** [comment]
Not sure I like sig_figs, I dont like writing underscores. I am fine with digits. I can change to that.

File: src/cmdstan/arguments/arg_output_precision.hpp [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    _validity = "integer >= 0 or -1 to use the default precision";
+    _default = "-1";
+    _default_value = -1;
+    _constrained = true;
+    _good_value = 8;
+    _bad_value = -2;
+    _value = _default_value;
+  }
+
+  bool is_valid(int value) { return value >= 0 || value == _default_value; }
```
Comment (mitzimorris) [inline src/cmdstan/arguments/arg_output_precision.hpp:22]: upper limit of 64-bit prececision is 15-17 - can we bound this at 18?
cf: https://en.wikipedia.org/wiki/Double-precision_floating-point_format

**Comment by mitzimorris** [comment]
>  I dont like writing underscores

it's naming convention used in the Stan code base - however regrettable (tooLateForCamelCase)

we need names to be concise, clear, and consistent. like many other sets of desiderata, they're in conflict.
given that this is a CmdStan user-facing thing, there's some precedent for `sig_figs`.  I admit it's not a great name either.

@seantalts, @jgabry, @imadmali - any thoughts?

**Comment by imadmali** [comment]
I think I've seen `digits` instead of `sig_figs` as args in other packages, so I'd prefer that. Also won't changing to `digits` break a lot code that was already using the `sig_figs` option? (Regarding camel case vs underscore I'm indifferent, so long as everything is consistent.)

**Comment by rok-cesnovar** [comment]
I did not mean it like I would prefer camel case. More that multiple words require an underscore (would prefer a dash but we have to stay consistent) so one word is preferred. 

But, I did not realize `stansummary` had `sig_figs` (or forgot) for the exact same thing as its being used here, so we should also go with `sig_figs` here. Apologize for the confusion. Fixed the code now.

**Review by mitzimorris — APPROVED** [review]
LGTM!

**Comment by mitzimorris** [comment]
(note: unit tests failing)

**Comment by mitzimorris** [comment]
needs doc:  https://github.com/stan-dev/docs/issues/273

**Pull request merged**